Userland/Libraries/LibWeb/WebAudio/AnalyserNode.cpp

```diff
@@ -0,0 +1,329 @@
+    //        need to implement up/down mixing somewhere
+    //        https://webaudio.github.io/web-audio-api/#channel-up-mixing-and-down-mixing
+    Vector<double> ret;
+    ret.ensure_capacity(m_fft_size);
+    return ret;
+}
+
+Vector<double> AnalyserNode::apply_a_blackman_window(Vector<double> const& x) const
+{
+    double const a = 0.16;
+    double const a0 = 0.5 * (1 - a);
+    double const a1 = 0.5;
+    double const a2 = a * 0.5;
+    unsigned long const N = m_fft_size;
+
+    auto w = [&](unsigned long n) {
+        return a0 - a1 * cos(2 * AK::Pi<double> * (double)n / (double)N) + a2 * cos(4 * AK::Pi<double> * (double)n / (double)N);
+    };
+
+    Vector<double> x_hat;
+    x_hat.ensure_capacity(m_fft_size);
+
+    // FIXME: Naive
+    for (unsigned long i = 0; i < m_fft_size; i++) {
+        x_hat[i] = x[i] * w(i);
+    };
+
+    return x_hat;
+}
+
+static Vector<double> apply_a_fourier_transform(Vector<double> const& input)
+{
+    dbgln("FIXME: Analyser node: implement apply a fourier transform");
+    auto ret = Vector<double>();
+    ret.ensure_capacity(input.size());
+    return ret;
+}
+
+Vector<double> AnalyserNode::smoothing_over_time(Vector<double> const& current_block)
+{
+    dbgln("FIXME: Analyser node: implement smoothing over time");
+    auto& X_hat_prev = m_previous_block;
+    auto tau = m_smoothing_time_constant;
+
+    // FIXME: Fourier transform
+    auto X = apply_a_fourier_transform(current_block);
+
+    // FIXME: Naive
+    Vector<double> X_hat;
+    X_hat.ensure_capacity(m_fft_size);
+    for (unsigned long i = 0; i < m_fft_size; i++) {
+        // FIMXE: Complex modulus on X[i]
+        X_hat[i] = tau * X_hat_prev[i] + (1 - tau) * abs(X[i]);
+    }
+
+    return X_hat;
+}
+
+// https://webaudio.github.io/web-audio-api/#conversion-to-db
+Vector<double> AnalyserNode::conversion_to_dB(Vector<double> const& X_hat) const
+{
+    Vector<double> Y;
+    Y.ensure_capacity(X_hat.size());
+    // FIXME: Naive
+    for (size_t i = 0; i < X_hat.size(); i++)
+        Y[i] = 20.0 * AK::log(X_hat[i]);
+
+    return Y;
+}
+
+// https://webaudio.github.io/web-audio-api/#current-frequency-data
+Vector<double> AnalyserNode::current_frequency_data()
+{
+    dbgln("FIXME: Analyser node: implement current frequency data");
+    // FIXME: 1. Compute the current time-domain data.
+    auto current_time_domain_dat = current_time_domain_data();
+
+    // FIXME: 2. Apply a Blackman window to the time domain input data.
+    auto blackman_windowed_input = apply_a_blackman_window(current_time_domain_dat);
+
+    // FIXME: 3. Apply a Fourier transform to the windowed time domain input data to get real and imaginary frequency data.
+    auto frequency_domain_dat = apply_a_fourier_transform(blackman_windowed_input);
+
+    // FIXME: 4. Smooth over time the frequency domain data.
+    auto smoothed_data = smoothing_over_time(frequency_domain_dat);
+
+    // FIXME: 5. Convert to dB.
+    return conversion_to_dB(smoothed_data);
```

Comment:
All the "FIXME:" parts can be removed, because the steps are implemented here, even if the functions they call have FIXMEs in.